### PR TITLE
 ITNFA-27: added biometric update support

### PR DESCRIPTION
**Pull request by pawelcieszko**

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/client/M2SysV105Client.java [hunk 1]
```diff
@@ -84,17 +88,21 @@ public BiometricSubject enroll(BiometricSubject subject) {
...
-        if (!results.isUpdateSuccess()) {
-            throw new M2SysBiometricsException("Unable to update fingerprints for: "
-                    + subject.getSubjectId());
+        String nationalId;
+        if (fingerScanStatus.isRegisteredNationally()) {
+            nationalId = fingerScanStatus.getNationalBiometricSubject().getSubjectId();
+            updateService.updateNationally(nationalId, fingerScan);
+        } else {
+            nationalId = nationalUuidGenerator.generate();
+            registrationService.registerNationally(nationalId, fingerScan);
```
Comment (dserkowski) [inline api/src/main/java/org/openmrs/module/m2sysbiometrics/client/M2SysV105Client.java]: The generate method uses also nationalClient, which could have i.e. incorrectly set credentials. If we want to persist this failure in the retry queue, we should keep it mind. I think we should move id generating to registrationService.

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/client/M2SysV105Client.java [hunk 2]
```diff
@@ -84,17 +88,21 @@ public BiometricSubject enroll(BiometricSubject subject) {
...
 
-        String response = localBioServerClient.update(subject.getSubjectId(), capture.getTemplateData());
-        M2SysResults results = XmlResultUtil.parse(response);
+        updateService.updateLocally(subject, fingerScan);
 
-        if (!results.isUpdateSuccess()) {
-            throw new M2SysBiometricsException("Unable to update fingerprints for: "
-                    + subject.getSubjectId());
+        String nationalId;
+        if (fingerScanStatus.isRegisteredNationally()) {
```
Comment (dserkowski) [inline api/src/main/java/org/openmrs/module/m2sysbiometrics/client/M2SysV105Client.java]: We should also check if national client is correctly configured

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/service/UpdateService.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package org.openmrs.module.m2sysbiometrics.service;
+
+import org.openmrs.module.m2sysbiometrics.model.M2SysCaptureResponse;
+import org.openmrs.module.registrationcore.api.biometrics.model.BiometricSubject;
+
+public interface UpdateService {
+
+    void updateLocally(BiometricSubject subject, M2SysCaptureResponse capture);
+
+    void updateNationally(String nationalId, M2SysCaptureResponse capture);
```
Comment (dserkowski) [inline api/src/main/java/org/openmrs/module/m2sysbiometrics/service/UpdateService.java]: capture -> fingerScan

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/service/UpdateService.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package org.openmrs.module.m2sysbiometrics.service;
+
+import org.openmrs.module.m2sysbiometrics.model.M2SysCaptureResponse;
+import org.openmrs.module.registrationcore.api.biometrics.model.BiometricSubject;
+
+public interface UpdateService {
+
+    void updateLocally(BiometricSubject subject, M2SysCaptureResponse capture);
```
Comment (dserkowski) [inline api/src/main/java/org/openmrs/module/m2sysbiometrics/service/UpdateService.java]: capture -> fingerScan

**Review by dserkowski — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/client/M2SysV105Client.java (hunk 1 above)
Comment (pawelcieszko) [reply api/src/main/java/org/openmrs/module/m2sysbiometrics/client/M2SysV105Client.java]: Right, in Oskar's PR it is moved so I won't change it for registration, but I will do it for update

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/client/M2SysV105Client.java (hunk 2 above)
Comment (pawelcieszko) [reply api/src/main/java/org/openmrs/module/m2sysbiometrics/client/M2SysV105Client.java]: right

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/service/UpdateService.java (hunk 3 above)
Comment (pawelcieszko) [reply api/src/main/java/org/openmrs/module/m2sysbiometrics/service/UpdateService.java]: ok

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/service/UpdateService.java (hunk 4 above)
Comment (pawelcieszko) [reply api/src/main/java/org/openmrs/module/m2sysbiometrics/service/UpdateService.java]: ok

**Review by dserkowski — APPROVED** [review]

**Pull request merged**